Speed improvements for `Acl#setRule()`, reducing complexity from `O(N^2)` to `O(N)`

JacobBrownAustin opened the pull request:
* Adding $resourcesById member variable so that we don't have to walk tree each time setRule is called
* removed unnecessary array_merge from getChildResources() and where we call getChildResources()
* Adding phpbench and a benchmark to cover these performance improvements

Signed-off-by: Jacob Brown <jacbrown@adobe.com>

Fixing https://github.com/laminas/laminas-permissions-acl/issues/6
<!--
Fill in the relevant information below to help triage your issue.

Pick the target branch based on the following criteria:
  * Documentation improvement: default X.Y.z branch or the oldest support X.Y.z
  * Bugfix: default X.Y.z branch or the oldest support X.Y.z
  * QA improvement (additional tests, CS fixes, etc.) that does not change code
    behavior: default X.Y.z branch or the oldest support X.Y.z
  * New feature, or refactor of existing code: develop branch

You MUST provide a signoff in your commits for us to be able to accept your
patch; you can do this by providing either the --signoff or -s flag when using
"git commit". Please see the project contributing guide and
https://developercertificate.org for details.
-->

|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | no
| RFC           | no
| QA            | no
| Performance            | yes

### Description

<!--
Tell us about why this change is necessary:
- Are you fixing a bug or providing a failing unit test to demonstrate a bug?
  - How do you reproduce it?
  - What did you expect to happen?
  - What actually happened?
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you adding documentation?
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you providing a QA improvement (additional tests, CS fixes, etc.) that
  does not change behavior?
  - Explain why the changes are necessary
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you fixing a BC Break?
  - How do you reproduce it?
  - What was the previous behavior?
  - What is the current behavior?
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you adding something the library currently does not support?
  - Why should it be added?
  - What will it enable?
  - How will the code be used?
  - TARGET THE develop BRANCH

- Are you refactoring code?
  - Why do you feel the refactor is necessary?
  - What types of refactoring are you doing?
  - TARGET THE develop BRANCH
-->

JacobBrownAustin committed:
Speed improvements for Acl::setRule

* Adding $resourcesById member variable so that we don't have to walk tree each time setRule is called
* removed unnecessary array_merge from getChildResources() and where we call getChildResources()

Signed-off-by: Jacob Brown <jacob@gnu.org>

Ocramius reviewed (changes requested):
Please do check the introduced CI failures (reported on the diff).

As for the speed improvements: any kind of measurement available?

JacobBrownAustin commented:
>Please do check the introduced CI failures (reported on the diff).

I will update soon.

>As for the speed improvements: any kind of measurement available?

In Magento admin,  I was measuring Magento\Authorization\Model\Acl\Loader\Rule::applyPermissionsAccordingToRules which uses this class.
It went from about 100-400 ms to about 10-15 ms.  It depends on how many resources there are multiplied by how many admin roles (which allow() and deny() turn into rules).

So in my testing,  I was getting at least 10x faster.  I believe that the array_merge calls in previous code might be causing the time to be O(n²),  so it might have been even worse for more admin roles.

I'll see if I can make a quick script to benchmark more generically it if I have time to do so.

Ocramius commented:
If you want to set up proper benchmarking, then a benchmark like the ones in https://github.com/laminas/laminas-servicemanager/tree/f22969b8b0ea2d695aa832c8895be29622b9bd94/benchmarks would be best: it already measures performance over time, and @phpbench is getting bigger/better with time :)

JacobBrownAustin commented:
Github won't let me reply to the complaints about ResourceInterface, so I'll just add it here as comment.

>src/Acl.php:778:62: MixedArgument: Argument 1 of Laminas\Permissions\Acl\Acl::getRuleType cannot be Laminas\Permissions\Acl\Resource\ResourceInterface|mixed|null, expecting Laminas\Permissions\Acl\Resource\ResourceInterface|null (see https://psalm.dev/030)

This is not code that I have changed.  This is a problem with the argument's variable getting reused and the static analyzer not realizing this and also that $this->resources is an array of untyped arrays instead of an array of objects of a defined class.  If you look at line 738,  
```$resource        >src/Acl.php:778:62: MixedArgument: Argument 1 of Laminas\Permissions\Acl\Acl::getRuleType cannot be Laminas\Permissions\Acl\Resource\ResourceInterface|mixed|null, expecting Laminas\Permissions\Acl\Resource\ResourceInterface|null (see https://psalm.dev/030)

This is not code that I have changed.  This is a problem with the argument's variable getting reused and the static analyzer not realizing this and also that $this->resources is an array of untyped arrays instead of an array of objects of a defined class.  If you look at line 738,  
```$resource                = $this->getResource($resource);```
you will see that $resource should always either be `null` or an object of ResourceInterface.  (This is assuming that $this->resources doesn't have bugs elsewhere because, like I said,  it is an array of arrays instead of array of objects of a defined class.)

Refactoring the $this->resources it outside the scope of this pull request.
        = $this->getResource($resource);```
you will see that $resource should always either be `null` or an object of ResourceInterface.  (This is assuming that $this->resources doesn't have bugs elsewhere because, like I said,  it is an array of arrays instead of array of objects of a defined class.)

Refactoring the $this->resources it outside the scope of this pull request.

JacobBrownAustin committed:
fixing static test failures for pull request #44

Signed-off-by: Jacob Brown <jacob@gnu.org>

JacobBrownAustin committed:
Adding phpbench benchmark for Acl

Signed-off-by: Jacob Brown <jacob@gnu.org>

JacobBrownAustin commented:
I created a benchmark using phpbench as you suggested.  Thanks for sharing that,  I hadn't used it before.
For my system,  the benchmark takes 71.833ms with this pull request,  and 2.512s before the pull request, which shows almost 35 times improvement in speed.

Ocramius commented on benchmarks/AclBench.php:
```diff
@@ -0,0 +1,68 @@
+<?php
```
Let's make sure this directory, as well as the phpbench config, are listed as `export-ignore` in `.gitattributes`

Ocramius commented on benchmarks/AclBench.php:
```diff
@@ -0,0 +1,68 @@
+        $this->acl = new Acl();
+    }
+
+    public function benchSetService(): void
```
Overall OK test scenario, naming could be improved perhaps.

This is good as-is, but in future (also for your own benchmarks), remember to split scenarios and use-cases like you would do with a unit test.

This scenario does an "all at once", heh 😁

JacobBrownAustin replied:
haha! Sorry about method name I had copied one of the examples you linked to and forgot to rename it.
Yes, for this testing of the allow() and deny(),  it requires many resources and roles to be set up in the ACL ahead of time.  I had thought about looking into how to separate that into different parts,  but wasn't sure how to do that in phpbench and just thought a working benchmark would be good for now.

Ocramius commented on phpbench.json:
```diff
@@ -0,0 +1,6 @@
+{
```
Should be in `export-ignore`

Ocramius commented on src/Acl.php:
```diff
@@ -58,6 +55,14 @@ class Acl implements AclInterface
+     *
+     * @var (ResourceInterface|null)[]
+     */
+    protected $resourcesById = [null];
```
Since this is a new property, let's make it `private` to prevent abuse

Ocramius replied:
Also, we do this for speed, at the cost of memory usage: best to keep this concept hidden/internal 👍

Ocramius commented on src/Acl.php:
```diff
@@ -58,6 +55,14 @@ class Acl implements AclInterface
+     * Resources by resourceId plus a null element
+     * Used to speed up setRule()
+     *
+     * @var (ResourceInterface|null)[]
```
The key type should be documented here

Ocramius commented on src/Acl.php:
```diff
@@ -563,35 +568,29 @@ public function setRule(
-                }
-            } else {
+        /** var (ResourceInterface|null)[] */
+        $resourcesArray = [];
```
This naming needs some work 😁

Ocramius commented on src/Acl.php:
```diff
@@ -563,35 +568,29 @@ public function setRule(
-                    array_unshift($resources, null);
-                }
-            } else {
+        /** var (ResourceInterface|null)[] */
```
Let's specify key types too

Ocramius commented on src/Acl.php:
```diff
@@ -563,35 +568,29 @@ public function setRule(
+                    $resourcesArray[] = null;
+                }
             }
+            unset($resources);
```
Is this `unset()` even necessary?

JacobBrownAustin replied:
It was already in there,  and I didn't want to increase memory requirements of this method when I changed the variable names to fix the static test error.  
(The static test error was complaining about that it thought that $resources could be a string because that was one of the optional parameter types,  so I created a new variable part way through the method and then got rid of the temporary variable.)

Ocramius replied:
Eh, it woud've been GC'd anyway, but let's leave it here for now :)

Ocramius commented on src/Acl.php:
```diff
@@ -675,21 +674,17 @@ public function setRule(
+     * @param (ResourceInterface|null)[] $return
+     * @return (ResourceInterface|null)[]
      */
-    protected function getChildResources(ResourceInterface $resource)
```
Uhh, we should really not mess with `protected` methods 🤔

They are very BC prone (mistake in the original design of the library, which is very old), so we should refrain from changing behaviour.

It would be best to make a new method instead, and deprecating the old one 🤔

JacobBrownAustin committed:
making requested changes for pull request #44

Signed-off-by: Jacob Brown <jacob@gnu.org>

JacobBrownAustin commented:
```Continuous Integration / ci / QA Checks (PHPCodeSniffer [8.0, locked], ubuntu-latest, laminas/laminas-continuous-integration-ac... (pull_request) Failing after 25s```

> Error: Incorrect annotations group.

I don't know what this means.  Do we just need to restart this test,  or does something in my pull request need adjusting?

Ocramius commented:
@JacobBrownAustin the `Incorrect annotations group` is something fixed by running `./vendor/bin/phpcbf`: the actual violation is visible in the PR diff, and it is complaining about the order of some annotations in a docblock :)

JacobBrownAustin commented:
Thanks for the suggestions.  I've made all requested changes and checks are all passing again.

Ocramius reviewed (approved):
Most excellent, thanks @JacobBrownAustin! 

Cutting a release now :)
